src/Juvix/Compiler/Backend/Geb/Analysis/TypeChecking.hs

```diff
@@ -0,0 +1,283 @@
+module Juvix.Compiler.Backend.Geb.Analysis.TypeChecking
+  ( module Juvix.Compiler.Backend.Geb.Analysis.TypeChecking,
+    module Juvix.Compiler.Backend.Geb.Analysis.TypeChecking.Data.Types,
+    module Juvix.Compiler.Backend.Geb.Analysis.TypeChecking.Error,
+  )
+where
+
+import Juvix.Compiler.Backend.Geb.Analysis.TypeChecking.Data.Types
+import Juvix.Compiler.Backend.Geb.Analysis.TypeChecking.Error
+import Juvix.Compiler.Backend.Geb.Data.Context as Context
+import Juvix.Compiler.Backend.Geb.Extra
+import Juvix.Compiler.Backend.Geb.Language
+
+check' :: Member (Error CheckingError) r => TypedMorphism -> Sem r TypedMorphism
+check' tyMorph = do
+  runReader defaultInferenceEnv $ check (tyMorph ^. typedMorphism) (tyMorph ^. typedMorphismObject)
+  return tyMorph
+
+check :: Members '[Reader InferenceEnv, Error CheckingError] r => Morphism -> Object -> Sem r ()
+check morph obj' = do
+  obj <-
+    runReader
+      (defaultInferenceEnv {_inferenceEnvTypeInfo = Just obj'})
+      (inferObject morph)
+  unless
+    (obj == obj')
+    ( throw $
+        CheckingErrorTypeMismatch
+          TypeMismatch
+            { _typeMismatchMorphism = morph,
+              _typeMismatchExpected = obj,
+              _typeMismatchActual = obj
+            }
+    )
+
+checkSameType :: InferEffects r => [Morphism] -> Sem r ()
+checkSameType = \case
+  [] -> return ()
+  (x : xs) -> do
+    obj <- inferObject x
+    checkListSameType xs obj
+
+checkListSameType :: InferEffects r => [Morphism] -> Object -> Sem r ()
+checkListSameType morphs obj = mapM_ (`check` obj) morphs
+
+inferObject' :: Morphism -> Either CheckingError Object
+inferObject' = run . runError . runReader defaultInferenceEnv . inferObject
+
+type InferEffects r = Members '[Reader InferenceEnv, Error CheckingError] r
+
+inferObject ::
+  Members '[Reader InferenceEnv, Error CheckingError] r =>
+  Morphism ->
+  Sem r Object
+inferObject = \case
+  MorphismUnit -> return ObjectTerminal
+  MorphismInteger {} -> return ObjectInteger
+  MorphismAbsurd x -> inferObject x
```

Comment:
This is incorrect. The type depends on the context, and cannot be inferred just based on the Absurd term if it's not provided in the term (see my comment in https://github.com/anoma/geb/issues/53)